cocotb/clock.py

```diff
@@ -165,3 +168,134 @@ async def start(self, cycles=None, start_high=True):
 
     def __str__(self):
         return type(self).__qualname__ + "(%3.1f MHz)" % self.frequency
+
+
+class SimClock(BaseClock):
+    r"""High performance, configurable simulator clock driver.
+
+    This class is a wrapper for a C++ layer clock driver with methods for control.
+    Use :meth:`start` to start the clock driver and :meth:`stop` to stop it.
+    The :attr:`is_running` property returns whether the driver is active.
+
+    Example:
+
+    .. code-block:: python
+
+        # defaults to 50:50 duty cycle
+        c = SimClock(dut.clk, 100, units='ns')
+        c.start()
+
+    Args:
+        signal: The clock pin/signal to be driven.
+        period (int, float, 2-tuple): The clock period. Must convert to an even number of
+            timesteps. Provide a 2-tuple as (high_time, low_time) to set the duty cycle manually.
+        units (str, optional): One of
+            ``'step'``, ``'fs'``, ``'ps'``, ``'ns'``, ``'us'``, ``'ms'``, ``'sec'``.
+            When *units* is ``'step'`` (the default),
+            the timestep is determined by the simulator (see :make:var:`COCOTB_HDL_TIMEPRECISION`).
+        duty_cycle (float, optional): The clock duty cycle as a float in [0, 1]. Default is 0.5.
```

Comment:
Is this a closed interval by design, i.e. `0.0` would produce a constant-0 (or, in case there is a `start_value` later, a constant-`start_value`)?

Reply:
I'm not sure if we should enforce at least 1 time step for high/low parts.
Icarus and Verilator work with 0/1, where you just get a 0-time glitch every period interval. The clock still toggles, it just toggles back in the same time step. Looking at the CI, Riviera may be having issues with that.

Reply:
Haha, I can already see someone trying to misuse that as an equivalent to a Verilog event. :)